sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/Repeater.java

```diff
@@ -65,28 +60,75 @@ static <InputT, OutputT> Builder<InputT, OutputT> builder() {
    */
   abstract BackOff getBackOff();
 
+  /** Counts invocations of {@link Caller#call}. */
+  abstract @Nullable Counter getCallCounter();
+
+  /** Convenience setter for {@link Builder#setCallCounter}. */
+  Repeater<InputT, OutputT> withCallCounter(@Nullable Counter value) {
+    if (value == null) {
+      return this;
+    }
+    return toBuilder().setCallCounter(checkStateNotNull(value)).build();
+  }
+
+  /** Counts invocations of {@link BackOff#nextBackOffMillis}. */
+  abstract @Nullable Counter getBackoffCounter();
+
+  /** Convenience setter for {@link Builder#setBackoffCounter}. */
+  Repeater<InputT, OutputT> withBackoffCounter(@Nullable Counter value) {
+    if (value == null) {
+      return this;
+    }
+    return toBuilder().setBackoffCounter(checkStateNotNull(value)).build();
+  }
+
+  /** Counts invocations of {@link Sleeper#sleep}. */
+  abstract @Nullable Counter getSleeperCounter();
+
+  /** Convenience setter for {@link Builder#setSleeperCounter}. */
+  Repeater<InputT, OutputT> withSleeperCounter(@Nullable Counter value) {
+    if (value == null) {
+      return this;
+    }
+    return toBuilder().setSleeperCounter(checkStateNotNull(value)).build();
+  }
+
+  abstract Builder<InputT, OutputT> toBuilder();
+
   /**
    * Applies the {@link InputT} to the {@link ThrowableFunction}, returning the {@link OutputT} if
-   * successful. If the function throws an exception that {@link #REPEATABLE_ERROR_TYPES} contains,
-   * repeats the invocation after {@link Sleeper#sleep} for the amount of time reported by {@link
-   * BackOff#nextBackOffMillis}. Throws the latest encountered {@link UserCodeExecutionException}
-   * when {@link BackOff} reports a {@link BackOff#STOP}.
+   * successful. If the function throws an exception that {@link
+   * RequestResponseIO#REPEATABLE_ERROR_TYPES} contains, repeats the invocation after {@link
+   * Sleeper#sleep} for the amount of time reported by {@link BackOff#nextBackOffMillis}. Throws the
+   * latest encountered {@link UserCodeExecutionException} when {@link BackOff} reports a {@link
+   * BackOff#STOP}.
    */
   OutputT apply(InputT input) throws UserCodeExecutionException {
     Optional<UserCodeExecutionException> latestError = Optional.empty();
     long waitFor = 0L;
     while (waitFor != BackOff.STOP) {
       try {
-        getSleeper().sleep(waitFor);
+        if (waitFor > 0L) {
+          if (getSleeperCounter() != null) {
```

Comment:
this is slightly different than previous comment but still checks null twice. It occurs frequently enough so consider a simple helper function
```
void incIfPresent(@Nullable counter) {
  if (counter != null) {
    counter.inc()
  }
}
```
to simplify. This also applied to Call.java

Use Optional counters could be another solution. 

```
if(getCounter().isPresent()){
  getCounter().inc();
}
```

Reply:
Thank you so much @Abacn for this suggestion. This really helped clean up the code.